### PR TITLE
Test session reuse with TLSv1.2 and TLSv.13

### PR DESCRIPTION
**Pull request by yschimke**
Discussing on https://github.com/google/conscrypt/issues/985

Shows failing case with TLSv1.3

```
Expecting not blank but was:<"">
	at okhttp3.ConscryptTest.testSessionReuse(ConscryptTest.kt:110)
```

### CODE REVIEW

**Comment by yschimke** [comment]
Confirmed that TLSv1.3 session ticket reuse is working with JSSE 15.  Test output from

```
  JsseDebugLogging.enableJsseDebugLogging {
    if (it.type == JsseDebugLogging.JsseDebugMessage.Type.Handshake) {
      println(it)
    }
  }
```

```
Produced ClientHello handshake message
"ClientHello": {

Consuming ServerHello handshake message
"ServerHello": {

Consuming ChangeCipherSpec message

Consuming EncryptedExtensions handshake message
"EncryptedExtensions": [

Consuming server Certificate handshake message
"Certificate": {
  "certificate_request_context": "",

Consuming CertificateVerify handshake message
"CertificateVerify": {
  "signature algorithm": ecdsa_secp256r1_sha256

Consuming server Finished handshake message
"Finished": {
  "verify data": {
    0000: 2F 1C 47 77 1B 76 5E 87   E1 0F 9B 9A 67 0C E3 1B  /.Gw.v^.....g...
    0010: D4 18 BE A9 48 E4 1E 6F   E7 DC 4F 06 B5 25 28 AE  ....H..o..O..%(.
  }'}

Produced client Finished handshake message
"Finished": {
  "verify data": {
    0000: 5D EE 37 0E 1F B3 4F 82   D8 6A 00 FF E3 DE D2 59  ].7...O..j.....Y
    0010: 6C 07 AD F7 37 9E AB 8A   CD 60 7A 18 59 AB C5 17  l...7....`z.Y...
  }'}

Consuming NewSessionTicket message
"NewSessionTicket": {
  "ticket_lifetime"      : "172,800",






Produced ClientHello handshake message
"ClientHello": {
  "client version"      : "TLSv1.2",
  "random"              : "74 78 7A 7B E6 47 13 C6 25 01 13 1B C1 C2 68 FA 2E F1 9B DA 4A 9B DE 0F 30 25 D7 19 77 85 B2 A9",
  "session id"          : "",
  "cipher suites"       : "[TLS_AES_256_GCM_SHA384(0x1302), TLS_AES_128_GCM_SHA256(0x1301), TLS_CHACHA20_POLY1305_SHA256(0x1303), TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384(0xC02C), TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256(0xC02B), TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256(0xCCA9), TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384(0xC030), TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256(0xCCA8), TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256(0xC02F), TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA(0xC014), TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA(0xC013), TLS_RSA_WITH_AES_256_GCM_SHA384(0x009D), TLS_RSA_WITH_AES_128_GCM_SHA256(0x009C), TLS_RSA_WITH_AES_256_CBC_SHA(0x0035), TLS_RSA_WITH_AES_128_CBC_SHA(0x002F)]",
  "compression methods" : "00",
...
    "session_ticket (35)": {
        "ticket" : {
          0000: 4F B1 60 00 15 52 1A 45   54 64 9E E8 6A 19 49 85  O.`..R.ETd..j.I.
....          
          0080: A1 AA 97 62 
        }
    },
Consuming ServerHello handshake message
"ServerHello": {
  "server version"      : "TLSv1.2",
  "random"              : "DE 91 CF 4D AC 00 79 CC 99 E9 8A 52 83 1E 72 AF D5 8A 78 9C 5C CE 36 D4 01 F0 0D 23 A1 72 C7 FD",


Consuming EncryptedExtensions handshake message
"EncryptedExtensions": [
  "application_layer_protocol_negotiation (16)": {
    [h2]
  }
]
Consuming server Finished handshake message
"Finished": {
  "verify data": {
Produced client Finished handshake message
"Finished": {
Consuming NewSessionTicket message
"NewSessionTicket": {
  "ticket_lifetime"      : "172,800",    
```

**Comment by yschimke** [comment]
Just the Handshake messages

```
Found resumable session. Preparing PSK message.
Produced ClientHello handshake message
Consuming ServerHello handshake message
Negotiated protocol version: TLSv1.3
Resuming session: 
Using PSK to derive early secret
Consuming EncryptedExtensions handshake message
Consuming server Finished handshake message
Produced client Finished handshake message
Consuming NewSessionTicket message
```

**Comment by yschimke** [comment]
Debug helper https://github.com/square/okhttp/pull/6579

File: okhttp/src/test/java/okhttp3/ConscryptTest.kt [hunk 1]
```diff
@@ -15,34 +15,129 @@
...
+      .build()
+
+    var reuseSession = false
+
+    val sslContext = handshakeCertificates.sslContext()
+    val systemSslSocketFactory = sslContext.socketFactory
+    val sslSocketFactory = object : DelegatingSSLSocketFactory(systemSslSocketFactory) {
+      override fun configureSocket(sslSocket: SSLSocket): SSLSocket {
+        return sslSocket.apply {
+          if (reuseSession) {
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/ConscryptTest.kt:87]: I find this somewhat hard to follow ... capturing a mutable local. Maybe I’m too used to how it works in Java and not Kotlin!

File: okhttp/src/test/java/okhttp3/ConscryptTest.kt [hunk 2]
```diff
@@ -15,34 +15,129 @@
...
+        }
+      }
+    }
+
+    client = client.newBuilder().connectionSpecs(listOf(spec))
+      .eventListenerFactory(clientTestRule.wrap(object : EventListener() {
+        override fun connectionAcquired(call: Call, connection: Connection) {
+          val sslSocket = connection.socket() as SSLSocket
+
+          sessionIds.add(sslSocket.session.id.toByteString().hex())
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/ConscryptTest.kt:99]: Omit the `.hex()` bit? Should still get a useful toString() if it fails.

File: okhttp/src/test/java/okhttp3/JSSETest.kt [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    // Force reuse. This appears flaky (30% of the time) even though sessions are reused.
+    // javax.net.ssl.SSLHandshakeException: No new session is allowed and no existing
+    // session can be resumed
+    // TODO: raise JDK bug.
+    // reuseSession = true
+
+    client.newCall(request).execute().use { response ->
+      assertEquals(200, response.code)
+    }
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/JSSETest.kt:193]: Extract a helper function that makes two connections in sequence and returns the session IDs? DRY this up.

**Review by swankjesse — APPROVED** [review]

File: okhttp/src/test/java/okhttp3/ConscryptTest.kt (hunk 2 above)
Comment (yschimke) [reply okhttp/src/test/java/okhttp3/ConscryptTest.kt:99]: This is awkward without assertThat helpers for Okio.  Like isEmpty

File: okhttp/src/test/java/okhttp3/ConscryptTest.kt (hunk 1 above)
Comment (yschimke) [reply okhttp/src/test/java/okhttp3/ConscryptTest.kt:87]: I'd argue that it's the most natural way to express it in Kotlin test code.  Creating additional one off fake classes with these features seems like a step back. I've used things like an AtomicBoolean to allow a mutable capture of a local variable in Java.

And with things like inlined functions making it less clear when you look at the call site whether a DSL is actually all in the same scope or has function calls.

File: okhttp/src/test/java/okhttp3/JSSETest.kt (hunk 3 above)
Comment (yschimke) [reply okhttp/src/test/java/okhttp3/JSSETest.kt:193]: Made a separate test instead

**Comment by yschimke** [comment]
Second connection working

```
Consuming ServerHello handshake message
"ServerHello": {
  "server version"      : "TLSv1.2",
  "random"              : "58 14 CC FF E3 30 5C F9 13 A0 AA 45 72 58 DE 4B 17 F8 52 0B 2E EE 63 04 44 4F 57 4E 47 52 44 01",
  "session id"          : "C9 9F 47 05 44 54 1F 18 5D 54 D6 86 94 2D C1 0C FA 92 4A 70 FF BA 2D 08 28 62 7E 30 CD 93 E6 92",
  "cipher suite"        : "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384(0xC02C)",
...
Produced ClientHello handshake message
"ClientHello": {
  "client version"      : "TLSv1.2",
  "random"              : "8D 65 BF C5 BA 6D 18 B8 9A 73 15 31 FA B0 E5 EE 28 E3 21 8E 12 28 05 F2 40 31 A2 94 F7 F1 D7 EE",
  "session id"          : "C9 9F 47 05 44 54 1F 18 5D 54 D6 86 94 2D C1 0C FA 92 4A 70 FF BA 2D 08 28 62 7E 30 CD 93 E6 92",
  "cipher suites"       : "[TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384(0xC02C), 
```

Second connection failing

```
Consuming ServerHello handshake message
"ServerHello": {
  "server version"      : "TLSv1.2",
  "random"              : "11 C1 1C 07 B4 C5 24 B6 42 B4 B3 4F 82 A5 2E D9 92 EB D5 F6 59 FC 8F C1 44 4F 57 4E 47 52 44 01",
  "session id"          : "91 60 74 EB 98 8D 60 5C 1D FB B6 6C EA F0 35 60 5E C0 8C E4 4F F3 44 94 B7 08 85 9F B5 E6 C1 AB",
  "cipher suite"        : "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384(0xC02C)",
...
Produced ClientHello handshake message
"ClientHello": {
  "client version"      : "TLSv1.2",
  "random"              : "40 96 82 BA F3 E8 39 16 67 6F A2 CA 1F 82 A2 F5 38 DC DF 4E BF 77 1C 58 D5 8C B7 C3 17 99 81 F5",
  "session id"          : "",
  "cipher suites"       : "[TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384(0xC02C), 
```
